### DISCUSSION TITLE
show unanswered questions at the end of timeline

### DISCUSSION THREAD

**Discussion by xliu382** [jspsych/jsPsych#1197] [Q&A] [answered]
Hi,

I'm trying to conduct an experiment where people need to click on one of four options to answer each question. I used `html-button-response` for this and set the `trial_duration` to be `6000`. 

After seeing all questions, they should see missed questions again in the end (i.e., questions where they did not click on any one of the four buttons due to time-out) and make response (if not answer, show again, until answered). I guess I should use `loop_function` and `jsPsych.addNodeToEndOfTimeline` and probably `jsPsych.pluginAPI` to realize this? But I cannot figure out how to structure them properly...  Thank you!

    /* create timeline */
    var timeline = []

    var questions = ['Q1', 'Q2', 'Q3', 'Q4', 'Q5', 'Q6'];
    var answers = ['A1', 'A2', 'A3','A4', 'A5', 'A6'];
    var s1 = answers.map((answer,i) => ({answer, question: questions[i]}));
    var s2 = jsPsych.randomization.sampleWithoutReplacement(s1, 4);
    var s2q = s2.map(function(a) {return a.question;});

    function addRespFromButtonScale(data, scale_name) {
        data.scale = scale_name;
        data.varname = scale_name + data.i;
        data.response = parseInt(data.button_pressed) + 1; // raw: 0, 1, 2, ...
    }

    var round2_options = ["I am sure I KNOW the answer", "I might KNOW the answer", "I might need to SEARCH the answer", "I am sure I need to SEARCH the answer"];

    var round2_trial = {
        timeline_variables: [
            {data: s2[0], q: s2q[0]},
            {data: s2[1], q: s2q[1]},
            {data: s2[2], q: s2q[2]},
            {data: s2[3], q: s2q[3]},
        ],
        timeline: [ 
            {type: "html-button-response",
            data: jsPsych.timelineVariable("data"),
            stimulus: jsPsych.timelineVariable("q"),
            choices: ["I am sure I KNOW the answer", "I might KNOW the answer", "I might need to SEARCH the answer", "I am sure I need to SEARCH the answer"],
            button_html: "<p><button class='jspsych-btn' style='font-size: 16px; font-weight: normal'>%choice%</button></p>",
            trial_duration: 4000,
            on_finish: function(data) {
                addRespFromButtonScale(data, "questions");
                jsPsych.pauseExperiment();
                jsPsych.addNodeToEndOfTimeline ({
                    loop_function: function(data) {
                        if (jsPsych.pluginAPI.convertButtonCharacterToButtonCode(null) == data.values()[0].button_press) {
                            return true;
                        } else {
                            return false;
                        }
                    }
                }, jsPsych.resumeExperiment)
            } },
            {type: 'html-keyboard-response',
            stimulus: '<div style="font-size:50px;">Please get ready for next question.</div>',
            choices: jsPsych.NO_KEYS,
            trial_duration: 1000
            },
        ],
        randomize_order: true
    };

    timeline.push(round2_trial);`

**Comment by becky-gilbert** [comment, marked as answer]
Hi @xliu382, I'm not sure if this is the best/easiest solution, but one way to do this is to create an empty array to store any unanswered questions. Then, in the trial's `on_finish` function, you can check to see if there was a response, and if not then add that question to your array. To loop back over any unanswered questions, you can create a looping timeline that presents the first question in this array, and then removes it from the array after it was presented again, and continues until there are no more unanswered questions. This looping timeline would need to go into a conditional timeline that checks if there are _any_ unanswered questions in the array, because you only want to run the loop if there is at least one unanswered question.
```javascript
var unanswered_questions = [];

var round2_trial = {
  timeline_variables: [
    // ...
  ],
  timeline: [
    // in your first trial, where you collect the response:
    on_finish: function(data) {
      if (data.button_pressed == null) {
        // add this question to the array of unanswered questions
        unanswered_questions.push(data.stimulus)
      }
    } 
    // the rest of your timeline ...
};

timeline.push(round2_trial);

var unanswered_trial = {
    type: "html-button-response",
    stimulus: function() {
        // this needs to be a function because the contents of the unanswered_questions array change during the task
        return unanswered_questions[0];
    },
    choices: ["I am sure I KNOW the answer", "I might KNOW the answer", "I might need to SEARCH the answer", "I am sure I need to SEARCH the answer"],
    button_html: "<p><button class='jspsych-btn' style='font-size: 16px; font-weight: normal'>%choice%</button></p>",
    trial_duration: 4000
};
var get_ready = {
    type: 'html-keyboard-response',
    stimulus: '<div style="font-size:50px;">Please get ready for next question.</div>',
    choices: jsPsych.NO_KEYS,
    trial_duration: 1000
};
var unanswered_loop = {
    timeline: [unanswered_trial, get_ready],
    loop_function: function() {
        // remove the first item from the array, which is the question that was just presented
        unanswered_questions.shift();
        if (unanswered_questions.length == 0) {
            return false;
        } else {
            return true;
        }
    }
};
var unanswered_conditional = {
    timeline: [unanswered_loop],
    conditional_function: function() {
        // only run the unanswered_loop if there were any unanswered questions
        // if there were no unanswered questions then this whole thing will be skipped
        if (unanswered_questions.length > 0) {
            return true;
        } else {
            return false;
        }
    }
}
timeline.push(unanswered_conditional);
```

**Reply by xliu382** [reply]
Thank you so much! I tried it and it worked well. However, it seemed like if I didn't answer the same question on a second time, the question stopped showing up on a third time. What can I change to make the looping continue until there is no answered question?

**Reply by becky-gilbert** [reply]
Yes, I think you can do that by modifying the `unanswered_trial` so that, if the question is left unanswered again, it's added back onto the `unanswered_questions` array. And the loop will just continue until the `unanswered_questions` array is empty. You can do that using the same method as in your first presentation of the questions, by adding an `on_finish` function to the `unanswered_trial` that checks if a response was made, and if not, adds the question back on to the array:
```javascript
var unanswered_trial = {
  // ...other parameters...
  on_finish: function(data) {
    if (data.button_pressed == null) {
      // add this question to the array of unanswered questions
      unanswered_questions.push(data.stimulus)
    }
  }
} 
```
Does that work?

**Reply by xliu382** [reply]
Yes it worked! Thank you so much :)

**Reply by becky-gilbert** [reply]
Great! And thanks for the update 😃
